consensus/istanbul/backend/engine.go

```diff
@@ -921,10 +921,67 @@ func (sb *backend) snapshot(chain consensus.ChainReader, number uint64, hash com
 		logger.Trace("Stored voting snapshot to disk", "number", snap.Number, "hash", snap.Hash)
 	}
 
+	sb.regen(chain, headers)
+
 	sb.recents.Add(snap.Hash, snap)
 	return snap, err
 }
 
+// regen commits snapshot data to database
+// If the last header number is larger than the last snapshot number, regenerate handler gets triggered.
```

Comment:
```suggestion
// regen is triggered if there is any checkpoint block in the `headers`.
// For each checkpoint block, this function verifies the existence of its snapshot in DB and stores one if missing.
```

I'm not sure how to read the diagram.. Does SI mean snapshot interval? Does it mean the last snapshot is in between header1 and headerN?

Reply:
SI = snapshot interval
last snapshot = latest snapshot interval number for the given header range (i.e., `to-(to%uint64(params.CheckpointInterval))`) 

```
 Triggered:
 |   ^                          ^                          ^                          ^  ...|
     SI                 SI*(last snapshot)                 SI                         SI
       			   | header1, .. headerN |
 Not triggered: (Guaranteed SI* was committed before )
 |   ^                          ^                          ^                          ^  ...|
     SI                 SI*(last snapshot)                 SI                         SI
                                  | header1, .. headerN |
```

Still uninterpretable?

Reply:
The last line was missing. Now i get it.

Reply:
Updated.